### PR TITLE
[4.0][com_installer] convert to prepared statement

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes

use prepared statement for SQL

### Testing Instructions
test com_installer

### Expected result
work as before

### Actual result
N/A

### CODE REVIEW

File: administrator/components/com_installer/Helper/InstallerHelper.php [hunk 1]
```diff
@@ -33,8 +33,8 @@ public static function getExtensionTypes()
 	{
 		$db    = Factory::getDbo();
 		$query = $db->getQuery(true)
-			->select('DISTINCT type')
-			->from('#__extensions');
+			->select('DISTINCT '. $db->quoteName('type'))
```
Comment (zero-24) [inline administrator/components/com_installer/Helper/InstallerHelper.php]:
```suggestion
			->select('DISTINCT ' . $db->quoteName('type'))
```

File: administrator/components/com_installer/Model/ManageModel.php [hunk 2]
```diff
@@ -336,24 +338,31 @@ protected function getListQuery()
 			}
 			else
 			{
-				$query->where('protected = 0')
-					->where('enabled = ' . (int) $status);
+				$status = (int) $status;
+				$query->where($db->quoteName('protected') .' = 0')
```
Comment (zero-24) [inline administrator/components/com_installer/Model/ManageModel.php]:
```suggestion
				$query->where($db->quoteName('protected') . ' = 0')
```

File: administrator/components/com_installer/Model/DatabaseModel.php [hunk 3]
```diff
@@ -350,30 +346,39 @@ protected function getListQuery()
...
+			$query->where($db->quoteName('extensions.type') . ' = :type')
+				->bind(':type', $type);
 		}
 
 		if ($clientId != '')
 		{
-			$query->where($db->quoteName('extensions.client_id') . ' = ' . (int) $clientId);
+			$clientId = (int) $clientId;
+			$query->where($db->quoteName('extensions.client_id') . ' = :clientid')
+				->bind(':clientid', $clientid, ParameterType::INTEGER);
```
Comment (Quy) [inline administrator/components/com_installer/Model/DatabaseModel.php]:
```suggestion
				->bind(':clientid', $clientId, ParameterType::INTEGER);
```

File: administrator/components/com_installer/Model/DatabaseModel.php [hunk 4]
```diff
@@ -350,30 +346,39 @@ protected function getListQuery()
...
 		}
 
 		if ($extensionId != '')
 		{
-			$query->where($db->quoteName('extensions.extension_id') . ' = ' . (int) $extensionId);
+			$extensionId = (int) $extensionId;
+			$query->where($db->quoteName('extensions.extension_id') . ' = :extensionid')
+				->bind(':extensionid', $extensionid, ParameterType::INTEGER);
```
Comment (Quy) [inline administrator/components/com_installer/Model/DatabaseModel.php]:
```suggestion
				->bind(':extensionid', $extensionId, ParameterType::INTEGER);
```

File: administrator/components/com_installer/Model/UpdateModel.php [hunk 5]
```diff
@@ -142,21 +156,29 @@ protected function getListQuery()
...
+					$uid = (int) substr($search, 3);
+					$query->where($db->quoteName('u.update_id') . ' = :uid')
+						->bind(':uid', $suid, ParameterType::INTEGER);
 				}
 				else
 				{
-					$query->where($db->quoteName('u.name') . ' LIKE ' . $db->quote('%' . str_replace(' ', '%', $db->escape(trim($search), true)) . '%'));
+					$search = '%' . str_replace(' ', '%', trim($search)) . '%';
+					$query->where($db->quoteName('u.name') . ' LIKE :search')
+						->bind(':search', $name);
```
Comment (Quy) [inline administrator/components/com_installer/Model/UpdateModel.php]:
```suggestion
						->bind(':search', $search);
```

File: administrator/components/com_installer/Model/UpdatesitesModel.php [hunk 6]
```diff
@@ -158,7 +159,7 @@ public function delete($ids = array())
 		$query = $db->getQuery(true)
 			->select($db->quoteName(array('update_site_id', 'name')))
```
Comment (Quy) [inline administrator/components/com_installer/Model/UpdatesitesModel.php]:
```suggestion
			->select($db->quoteName(['update_site_id', 'name']))
```

**Comment by alikon** [comment]
this need a redo after #25553 merge

File: administrator/components/com_installer/Model/DatabaseModel.php [hunk 7]
```diff
@@ -350,30 +346,39 @@ protected function getListQuery()
...
-			$query->where($db->quoteName('extensions.extension_id') . ' = ' . (int) $extensionId);
+			$extensionId = (int) $extensionId;
+			$query->where($db->quoteName('extensions.extension_id') . ' = :extensionid')
+				->bind(':extensionid', $extensionId, ParameterType::INTEGER);
 		}
 
 		if ($folder != '' && in_array($type, array('plugin', 'library', '')))
 		{
-			$query->where($db->quoteName('extensions.folder') . ' = ' . $db->quote($folder == '*' ? '' : $folder));
+			$folder == '*' ? '' : $folder;
```
Comment (Quy) [inline administrator/components/com_installer/Model/DatabaseModel.php]:
```suggestion
			$folder = $folder == '*' ? '' : $folder;
```

File: administrator/components/com_installer/Model/DatabaseModel.php [hunk 8]
```diff
@@ -417,11 +422,13 @@ protected function mergeSchemaCache($results)
...
-		$db    = $this->getDbo();
-		$query = $db->getQuery(true)
+		$db          = $this->getDbo();
+		$extensionId = (int) $extensionId;
+		$query       = $db->getQuery(true)
 			->select($db->quoteName('version_id'))
 			->from($db->quoteName('#__schemas'))
-			->where($db->quoteName('extension_id') . ' = ' . (int) $extensionId);
+			->where($db->quoteName('extension_id') . ' = :extensionid')
+			->bind(':extensionid', $extensionid, ParameterType::INTEGER);
```
Comment (Quy) [inline administrator/components/com_installer/Model/DatabaseModel.php]:
```suggestion
			->bind(':extensionid', $extensionId, ParameterType::INTEGER);
```

File: administrator/components/com_installer/Model/DiscoverModel.php [hunk 9]
```diff
@@ -104,17 +105,22 @@ protected function getListQuery()
...
-			$query->where($db->quoteName('client_id') . ' = ' . (int) $clientId);
+			$clientId = (int) $clientId;
+			$query->where($db->quoteName('client_id') . ' = :clientid')
+				->bind(':clientid', $clientId, ParameterType::INTEGER);
 		}
 
 		if ($folder != '' && in_array($type, array('plugin', 'library', '')))
 		{
-			$query->where($db->quoteName('folder') . ' = ' . $db->quote($folder == '*' ? '' : $folder));
+			$folder == '*' ? '' : $folder;
```
Comment (Quy) [inline administrator/components/com_installer/Model/DiscoverModel.php]:
```suggestion
			$folder = $folder == '*' ? '' : $folder;
```

File: administrator/components/com_installer/Model/ManageModel.php [hunk 10]
```diff
@@ -336,24 +338,31 @@ protected function getListQuery()
...
-			$query->where('client_id = ' . (int) $clientId);
+			$clientId = (int) $clientId;
+			$query->where($db->quoteName('client_id') . ' = :clientid')
+				->bind(':clientid', $clientId, ParameterType::INTEGER);
 		}
 
 		if ($folder !== '')
 		{
-			$query->where('folder = ' . $this->_db->quote($folder == '*' ? '' : $folder));
+			$folder == '*' ? '' : $folder;
```
Comment (Quy) [inline administrator/components/com_installer/Model/ManageModel.php]:
```suggestion
			$folder = $folder == '*' ? '' : $folder;
```

File: administrator/components/com_installer/Model/UpdateModel.php [hunk 11]
```diff
@@ -112,27 +117,36 @@ protected function getListQuery()
...
-			$query->where($db->quoteName('u.client_id') . ' = ' . (int) $clientId);
+			$clientId = (int) $clientId;
+			$query->where($db->quoteName('u.client_id') . ' = :clientid')
+				->bind(':clientid', $clientId, ParameterType::INTEGER);
 		}
 
 		if ($folder != '' && in_array($type, array('plugin', 'library', '')))
 		{
-			$query->where($db->quoteName('u.folder') . ' = ' . $db->quote($folder == '*' ? '' : $folder));
+			$folder == '*' ? '' : $folder;
```
Comment (Quy) [inline administrator/components/com_installer/Model/UpdateModel.php]:
```suggestion
			$folder = $folder == '*' ? '' : $folder;
```

File: administrator/components/com_installer/Model/UpdateModel.php [hunk 12]
```diff
@@ -142,21 +156,29 @@ protected function getListQuery()
...
+					$suid = (int) substr($search, 4);
+					$query->where($db->quoteName('u.update_site_id') . ' = :suid')
+						->bind(':suid', $suid, ParameterType::INTEGER);
 				}
 				elseif (stripos($search, 'id:') !== false)
 				{
-					$query->where($db->quoteName('u.update_id') . ' = ' . (int) substr($search, 3));
+					$uid = (int) substr($search, 3);
+					$query->where($db->quoteName('u.update_id') . ' = :uid')
+						->bind(':uid', $suid, ParameterType::INTEGER);
```
Comment (Quy) [inline administrator/components/com_installer/Model/UpdateModel.php]:
```suggestion
						->bind(':uid', $uid, ParameterType::INTEGER);
```

File: administrator/components/com_installer/Model/UpdatesitesModel.php [hunk 13]
```diff
@@ -528,23 +543,26 @@ protected function getListQuery()
...
 		if ($clientId !== null && $clientId !== '')
 		{
+			$clientId = (int) $clientId;
 			$query->where($db->quoteName('e.client_id') . ' = :clientId')
 				->bind(':clientId', $clientId, ParameterType::INTEGER);
 		}
 
 		if ($folder !== '' && in_array($type, ['plugin', 'library', ''], true))
 		{
+			$folder === '*' ? '' : $folder;
```
Comment (Quy) [inline administrator/components/com_installer/Model/UpdatesitesModel.php]:
```suggestion
			$folder = $folder === '*' ? '' : $folder;
```

File: administrator/components/com_installer/Model/DatabaseModel.php (hunk 7 above)
Comment (Quy) [inline administrator/components/com_installer/Model/DatabaseModel.php]:
```suggestion
			$folder = $folder === '*' ? '' : $folder;
```

File: administrator/components/com_installer/Model/DiscoverModel.php (hunk 9 above)
Comment (Quy) [inline administrator/components/com_installer/Model/DiscoverModel.php]:
```suggestion
			$folder = $folder === '*' ? '' : $folder;
```

File: administrator/components/com_installer/Model/ManageModel.php (hunk 10 above)
Comment (Quy) [inline administrator/components/com_installer/Model/ManageModel.php]:
```suggestion
			$folder = $folder === '*' ? '' : $folder;
```

File: administrator/components/com_installer/Model/UpdateModel.php (hunk 11 above)
Comment (Quy) [inline administrator/components/com_installer/Model/UpdateModel.php]:
```suggestion
			$folder = $folder === '*' ? '' : $folder;
```

**Comment by alikon** [comment]
conflict solved

File: administrator/components/com_installer/Model/UpdatesitesModel.php [hunk 14]
```diff
@@ -145,7 +145,7 @@ public function delete($ids = [])
 		$query = $db->getQuery(true)
 			->select($db->quoteName(['update_site_id', 'name']))
 			->from($db->quoteName('#__update_sites'))
-			->where($db->quoteName('update_site_id') . ' IN (' . implode(', ', $ids) . ')');
+			->whereIn($db->quoteName('update_site_id'), $ids);
```
Comment (wilsonge) [inline administrator/components/com_installer/Model/UpdatesitesModel.php]: I think this should be type'd to integer

File: administrator/components/com_installer/Model/ManageModel.php [hunk 15]
```diff
@@ -363,18 +372,18 @@ protected function getListQuery()
...
-				$elements[] = $this->getDbo()->quote($extension[1]);
+				$elements[] = $extension[1];
 			}
 
 			if ($elements)
 			{
 				if ($core === '1')
 				{
-					$query->where($this->getDbo()->quoteName('element') . ' IN (' . implode(',', $elements) . ')');
+					$query->whereIn($db->quoteName('element'), $elements);
```
Comment (wilsonge) [inline administrator/components/com_installer/Model/ManageModel.php]: This needs type'ing to string? `element` is type string and `whereIn` is integer by default

File: administrator/components/com_installer/Model/ManageModel.php [hunk 16]
```diff
@@ -363,18 +372,18 @@ protected function getListQuery()
...
 				}
 				elseif ($core === '0')
 				{
-					$query->where($this->getDbo()->quoteName('element') . ' NOT IN (' . implode(',', $elements) . ')');
+					$query->whereNotIn($db->quoteName('element'), $elements);
```
Comment (wilsonge) [inline administrator/components/com_installer/Model/ManageModel.php]: I think this is the same

**Comment by wilsonge** [comment]
two fixes required and this is good to go

File: administrator/components/com_installer/Model/ManageModel.php (hunk 15 above)
Comment (alikon) [reply administrator/components/com_installer/Model/ManageModel.php]: yes, done

File: administrator/components/com_installer/Model/ManageModel.php (hunk 16 above)
Comment (alikon) [reply administrator/components/com_installer/Model/ManageModel.php]: yes, done

**Comment by wilsonge** [comment]
I have tested this item :white_check_mark: successfully on 661b0be44f384b967d296c2b9c376e1ec2261420<br /><br />Code reviewed this successfully.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26085">issues.joomla.org/tracker/joomla-cms/26085</a>.</sub>

File: administrator/components/com_installer/Model/UpdatesitesModel.php [hunk 17]
```diff
@@ -373,13 +378,18 @@ public function rebuild(): void
 						$query = $db->getQuery(true)
 							->select($db->quoteName('extension_id'))
 							->from($db->quoteName('#__extensions'))
-							->where(
-								'(' . $db->quoteName('name') . ' = ' . $db->quote($manifest->name)
-								. ' OR ' . $db->quoteName('name') . ' = ' . $db->quote($manifest->packagename) . ')'
+							->extendWhere(
```
Comment (SharkyKZ) [inline administrator/components/com_installer/Model/UpdatesitesModel.php:382]: Move this lower so there's at least 1 `where()` call before `extendWhere()`.

Comment (alikon) [reply administrator/components/com_installer/Model/UpdatesitesModel.php:382]: good catch
moved

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 661b0be44f384b967d296c2b9c376e1ec2261420<br /><br />Conde review + brief test of installing language, installing extension, updating extension, managing extension update site's download key.

Error reporting set to maximum, no errors or warning or notices in PHP log, no errors or warning in MySQL log.

Tested with MySQL 5.7

Thanks to PR #26085 for the Null Files extension for testing download key and update.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26085">issues.joomla.org/tracker/joomla-cms/26085</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 661b0be44f384b967d296c2b9c376e1ec2261420<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26085">issues.joomla.org/tracker/joomla-cms/26085</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26085">issues.joomla.org/tracker/joomla-cms/26085</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**